libeos-updater-util/flatpak.c

```diff
@@ -0,0 +1,1411 @@
+  const gchar *app_name = NULL;
+  FlatpakRefKind kind;
+
+  if (!parse_flatpak_ref_from_entry (entry, &app_name, &kind, error))
+    return NULL;
+
+  return g_object_new (FLATPAK_TYPE_REMOTE_REF,
+                       "remote-name", "none",
+                       "name", app_name,
+                       "kind", kind,
+                       NULL);
+}
+
+static FlatpakRemoteRef *
+flatpak_remote_ref_from_action_entry (EosUpdaterUtilFlatpakRemoteRefActionType   action_type,
+                                      JsonObject                                *entry,
+                                      GError                                   **error)
+{
+  switch (action_type)
+    {
+      case EUU_FLATPAK_REMOTE_REF_ACTION_INSTALL:
+        return flatpak_remote_ref_from_install_action_entry (entry, error);
+      case EUU_FLATPAK_REMOTE_REF_ACTION_UNINSTALL:
+        return flatpak_remote_ref_from_uninstall_action_entry (entry, error);
+      default:
+        g_assert_not_reached ();
+    }
+}
+
+static FlatpakRemoteRefAction *
+flatpak_remote_ref_action_from_json_node (JsonNode *node,
+                                          GError **error)
+{
+  const gchar *action_type_str = NULL;
+  JsonObject *object = json_node_get_object (node);
+  g_autoptr(FlatpakRemoteRef) flatpak_remote_ref = NULL;
+  g_autoptr(GError) local_error = NULL;
+  JsonNode *serial_node = NULL;
+  gint64 serial64;
+  gint32 serial;
+  EosUpdaterUtilFlatpakRemoteRefActionType action_type;
+
+  if (object == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected entry node to be an object, but was %s", node_str);
+
+      return NULL;
+    }
+
+  action_type_str = json_object_get_string_member (object, "action");
+
+  if (action_type_str == NULL)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected an 'action' member in the autoinstall spec (at %s)", node_str);
+
+      return NULL;
+    }
+
+  if (!flatpak_remote_ref_action_type_parse (action_type_str, &action_type, error))
+    return NULL;
+
+  serial_node = json_object_get_member (object, "serial");
+  if (serial_node == NULL ||
+      !JSON_NODE_HOLDS_VALUE (serial_node) ||
+      json_node_get_value_type (serial_node) != G_TYPE_INT64)
+    {
+      g_autofree gchar *node_str = json_node_to_string (node);
+      g_set_error (error,
+                   EOS_UPDATER_ERROR,
+                   EOS_UPDATER_ERROR_MALFORMED_AUTOINSTALL_SPEC,
+                   "Expected a 'serial' member in the autoinstall spec (at %s)", node_str);
+      return NULL;
+    }
+
+  serial64 = json_node_get_int (serial_node);
+
+  if (serial64 > G_MAXINT32 || serial64 < G_MININT32)
```

Comment:
This restriction isn’t enforced in the `.schema.json` file. (It makes sense, but it should be added to the JSON Schema.)